hazelcast/src/test/java/com/hazelcast/internal/ascii/MemcachedRawTcpTest.java

```diff
@@ -0,0 +1,614 @@
+        assertEquals("STORED", receiveLine());
+
+        sendLine("get key 0 0 3");
+        assertEquals("VALUE key 0 6", receiveLine());
+        assertEquals("barfoo", receiveLine());
+        assertEquals("END", receiveLine());
+    }
+
+    @Test
+    public void testExpiration() throws Exception {
+        sendLine("set key 0 3 3");
+        sendLine("foo");
+        assertEquals("STORED", receiveLine());
+
+        assertTrueEventually(() -> {
+            sendLine("get key");
+            if (!receiveLine().equals("END")) {
+                receiveLine(); // foo
+                receiveLine(); // END
+                fail();
+            }
+        });
+    }
+
+    @Test
+    public void testSetGet_withLargeValue() throws Exception {
+        int capacity = 10000;
+        StringBuilder value = new StringBuilder(capacity);
+        while (value.length() < capacity) {
+            value.append(randomString());
+        }
+
+        sendLine("set key 0 0 " + value.length());
+        sendLine(value.toString());
+        assertEquals("STORED", receiveLine());
+
+        sendLine("get key");
+        assertEquals("VALUE key 0 " + value.length(), receiveLine());
+        assertEquals(value.toString(), receiveLine());
+    }
+
+    @Test
+    public void testBulkSetGet_withManyKeys() throws Exception {
+        int numberOfKeys = 1000;
+
+        Map<String, String> entries = new HashMap<>();
+        for (int i = 0; i < numberOfKeys; i++) {
+            String key = "key" + i;
+            String value = "value" + i;
+            entries.put(key, value);
+
+            sendLine(String.format("set %s 0 0 %d", key, value.length()));
+            sendLine(value);
+            assertEquals("STORED", receiveLine());
+        }
+
+        sendLine("get " + String.join(" ", entries.keySet()));
+
+        for (int i = 0; i < numberOfKeys; i++) {
+            Matcher matcher = valueHeaderPattern.matcher(receiveLine());
+            assertTrue(matcher.matches());
+
+            String key = matcher.group(1);
+            int flags = Integer.parseInt(matcher.group(2));
+            int length = Integer.parseInt(matcher.group(3));
+
+            assertEquals(0, flags);
+
+            String value = receiveLine();
+            assertEquals(length, value.length());
+            assertEquals(entries.get(key), value);
+
+            // remove the entry to not count it again
+            entries.remove(key);
+        }
+        assertEquals("END", receiveLine());
+    }
+
+    /**
+     * Sends a single request line to the server.
+     * Adds CRLF at the end and flushes.
+     * @param line line to send
+     */
+    private void sendLine(String line) {
+        writer.write(line + CRLF);
+        writer.flush();
+    }
+
+    /**
+     * Tries to receive a response line from the server,
+     * calls Assert.fail() if nothing was received in 1 second.
+     * @return received line
+     */
+    private String receiveLine() throws Exception {
```

Comment:
@yemreinci Why did you need a a timeout ? 
Can't we just make this a blocking call ?

Reply:
We can, and it would work. But if somehow, in the future, the implementation gets broken and starts to return an error for, say, a get call, the test will expect 2 lines but the server will send only one. And the test will hang indefinitely on CI or sth and that would be hard to debug.

Reply:
Also, while implementing it I didn't get it right in the first time, of course, and hit this timeout. So it was better to have tests that failed instead of ones that hang indefinitely. 
If you have a better suggestion on how to implement a receiveLine() with timeout, please say. This was the best I could do with my Java knowledge.

Reply:
Hmm, actually, there is `@Time(timeout = x)`. Would it be better if I get rid of the timeout logic from this method and put timeout on tests individually? I would have to put different timeouts for each each test though, because some of them take more time than the others. (the one that tests expiration, for example).

Reply:
I think we don't need any timeout on the tests. Think of this similar to any operation on the API. For example, we are not putting a timeout to tests because `IMap.get()` might hang forever. 
The worst case is that, a test build will timeout and we will find the problem.

Reply:
OK, removed the timeout logic in https://github.com/hazelcast/hazelcast/pull/19653/commits/d5de9ed6f98dc3ad3eb91f9f95cf641803660bf3